ruby-hardcoded-api/mfclient.rb

```diff
@@ -0,0 +1,224 @@
+      <request>
+        <service name="asset.set" session="#{@session_id}">
+          <args>
+            <id>#{id}</id>
+            <meta>
+              <mf-note>
+                <note>#{mf_note}</note>
+              </mf-note>
+            </meta>
+          </args>
+        </service>
+      </request>
+    XML_BODY
+    response_body = http_post(xml_request)
+  end
+
+  # Creates an empty file (no content) with the name provided
+  def create(namespace, filename)
+    xml_request = <<-XML_BODY
+      <request>
+        <service name="asset.create" session="#{@session_id}" data-out-min="0" data-out-max="0">
+          <args>
+            <name>#{filename}</name>
+            <namespace>#{namespace}</namespace>
+          </args>
+        </service>
+      </request>
+    XML_BODY
+    response_body = http_post(xml_request)
+  end
+
+  # Uploads a file to the given namespace
+  def upload(namespace, filename_fullpath)
+    filename = File.basename(filename_fullpath)
+    xml_request = <<-XML_BODY
+    <request>
+      <service name="asset.create" session="#{@session_id}">
+        <args>
+          <namespace create="True">#{namespace}</namespace>
+          <name>#{filename}</name>
+          <meta><mf-name><name>#{filename}</name></mf-name></meta>
+        </args>
+        <attachment></attachment>
+      </service>
+    </request>
+    XML_BODY
+    file_content = File.read(filename_fullpath)
+    response_body = http_post(xml_request, true, file_content)
+  end
+
+  private
+    def http_post(payload, mflux = false, file_content = nil)
+      url = @base_url
+      uri = URI.parse(url)
+      http = Net::HTTP.new(uri.host, uri.port)
+      if url.start_with?("https://")
+        http.use_ssl = true
+        http.verify_mode = OpenSSL::SSL::VERIFY_NONE
+      end
+      request = Net::HTTP::Post.new(url)
+
+      xml = @xml_declaration + payload
+      if mflux == false
+        request["Content-Type"] = "text/xml"
+        request.body = xml
+      else
+        # Here be dragons
+        # Requests are built different for this content-type
+        request["Content-Type"] = "application/mflux"
+        mflux_request = if file_content.nil?
+                          xml_separator(xml) + xml
+                        else
+                          xml_separator(xml) + xml + content_separator(file_content) + file_content
+                        end
+        request.body = mflux_request
+      end
+
+      response = http.request(request)
+      if response.content_type == "application/mflux"
+        # More dragons
+        # Horrible hack to extract the content
+        header = response.body[0..23]
+        metadata_size = length_from_header(header)
+        start_at = metadata_size + 24 + 24 + 2  # metadata_size + header + header + 2
+        response.body[start_at..]
+      else
+        response.body
+      end
+    end
+
+    def xml_separator(xml)
+      # 01 00 xx xx xx xx xx xx xx xx 00 00 00 01 yy yy
```

Comment:
MediaFlux uses 24-byte headers to separate metadata from content. I am not sure where they are documented, though.